File: cppProjects/logica/02_anagrama.cpp
/*
 * Escribe una función que reciba dos palabras (String) y retorne
 * verdadero o falso (Bool) según sean o no anagramas.
 * - Un Anagrama consiste en formar una palabra reordenando TODAS
 *   las letras de otra palabra inicial.
 * - NO hace falta comprobar que ambas palabras existan.
 * - Dos palabras exactamente iguales no son anagrama.
 */

#include <iostream>
#include <algorithm>
#include <string>
#include <cctype>

using namespace std;

bool sonAnagramas(const string &palabraUno, const string &palabraDos)
{
    // Dos palabras exactamente iguales no son anagramas
    if (palabraUno == palabraDos)
    {
        return false;
    }

    // Crear copias de las palabras para no modificar las originales
    string copiaPalabraUno = palabraUno;
    string copiaPalabraDos = palabraDos;

    // Ordenar las copias de las palabras
    sort(copiaPalabraUno.begin(), copiaPalabraUno.end());
    sort(copiaPalabraDos.begin(), copiaPalabraDos.end());

    // Comparar las palabras ordenadas
    return copiaPalabraUno == copiaPalabraDos;
}

int main()
{
    string palabraUno, palabraDos;
    cout << "Ingrese la primera palabra: ";
    cin >> palabraUno;
    cout << "Ingrese la segunda palabra: ";
    cin >> palabraDos;

    // De esta manera se convierten las cadenas de texto a minúsculas
    for (char &c : palabraUno)
    {
        c = tolower(c);
    }
    for (char &c : palabraDos)
    {
        c = tolower(c);
    }

    bool anagrama = sonAnagramas(palabraUno, palabraDos);
    if (anagrama)
    {
        cout << "Las palabras son anagramas." << endl;
    }
    else
    {
        cout << "Las palabras no son anagramas." << endl;
    }
    return 0;
}

File: pythonProjects/logica/02_anagrama.py
import os


def limpiar_consola():
    os.system("cls")


def ingresar_palabras():
    palabra_uno = input("Ingresa un Palabra: ").lower()
    palabra_dos = input("Ingresa otra palabra: ").lower()
    return palabra_uno, palabra_dos


def anagrama():
    limpiar_consola()
    palabra_uno, palabra_dos = ingresar_palabras()

    if palabra_uno == palabra_dos:
        return print("No es un anagrama")
    # La función set() en Python se utiliza para crear un conjunto, que es una colección desordenada de elementos únicos.
    if set(palabra_uno) == set(palabra_dos):
        return print("Es un anagrama")
    else:
        return print("No es un anagrama")


if __name__ == "__main__":
    anagrama()
